src/AMAMM.sol

```diff
@@ -2,69 +2,200 @@
-                afterRemoveLiquidityReturnDelta: false
-            });
+    /// -----------------------------------------------------------------------
+    /// Library usage
+    /// -----------------------------------------------------------------------
+
+    using SafeCastLib for *;
+    using FixedPointMathLib for *;
+
+    /// -----------------------------------------------------------------------
+    /// Constants
+    /// -----------------------------------------------------------------------
+
+    function K(PoolId) public view virtual returns (uint40) {
+        return 24;
+    }
+
+    function EPOCH_SIZE(PoolId) public view virtual returns (uint256) {
+        return 1 hours;
+    }
+
+    function MIN_BID_MULTIPLIER(PoolId) public view virtual returns (uint256) {
+        return 1.1e18;
+    }
+
+    /// -----------------------------------------------------------------------
+    /// Storage variables
+    /// -----------------------------------------------------------------------
+
+    mapping(PoolId id => uint40) internal _lastUpdatedEpoch;
+    mapping(address deposits => uint256) public _userBalance;
+    mapping(Currency currency => uint256) internal _totalFees;
+    mapping(PoolId id => mapping(uint40 => Bid)) public poolEpochManager;
+    mapping(address manager => mapping(Currency currency => uint256)) internal _fees;
+    mapping(PoolId id => address) internal _bidToken;
+
+    /// -----------------------------------------------------------------------
+    /// Getter actions
+    /// -----------------------------------------------------------------------
+
+    function getLastManager(PoolId id, uint40 targetEpoch) public view returns (Bid memory) {
+        if (_lastUpdatedEpoch[id] > targetEpoch) {
+            if (_lastUpdatedEpoch[id] - targetEpoch <= K(id)) return poolEpochManager[id][_lastUpdatedEpoch[id]];
+            else return poolEpochManager[id][targetEpoch];
+        } else {
+            if (targetEpoch - _lastUpdatedEpoch[id] <= K(id)) {
+                return poolEpochManager[id][_lastUpdatedEpoch[id]];
+            } else {
+                return poolEpochManager[id][targetEpoch];
+            }
+        }
+    }
+
+    /// -----------------------------------------------------------------------
+    /// Bidder actions
+    /// -----------------------------------------------------------------------
+    /// @inheritdoc IAmAmm
+    function bid(PoolId id, bytes7 payload, uint128 rent, uint40 _epoch) external virtual override {
+        /// -----------------------------------------------------------------------
+        /// Validation
+        /// -----------------------------------------------------------------------
+
+        address msgSender = LibMulticaller.senderOrSigner();
+        if (_epoch > _getEpoch(id, block.timestamp) + K(id)) {
+            revert AmAmm__BidOutOfBounds();
+        }
+
+        uint128 deposit = depositToken(id, msgSender, rent);
```

Comment:
```suggestion
        depositToken(id, msgSender, rent);
```
Don't need the return value and validation as it's already been done at `depositToken`